### PR TITLE
8321180: Condition for non-latin1 string size too large exception is off by one

### PR DESCRIPTION
**Pull request by RogerRiggs**
In the compact string implementation of non-latin1 (UTF16) strings the length is constrained by VM implementation limit on the size a byte array that can be allocated. To produce a useful exception the implementation checks the string size against the maximum byte array size. The exception message is correct
```
"UTF16 String size is " + len + ", should be less than or equal to " + MAX_LENGTH
```
The code should match the message, otherwise the exception thrown is: 
```
java.lang.OutOfMemoryError: Requested array size exceeds VM limit 
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321180](https://bugs.openjdk.org/browse/JDK-8321180): Condition for non-latin1 string size too large exception is off by one (**Bug** - P4)

### Reviewers
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17008/head:pull/17008` \
`$ git checkout pull/17008`

Update a local copy of the PR: \
`$ git checkout pull/17008` \
`$ git pull https://git.openjdk.org/jdk.git pull/17008/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17008`

View PR using the GUI difftool: \
`$ git pr show -t 17008`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17008.diff">https://git.openjdk.org/jdk/pull/17008.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17008#issuecomment-1843861134)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/StringUTF16.java [hunk 1]
```diff
@@ -59,9 +54,9 @@ public static int newBytesLength(int len) {
         if (len < 0) {
             throw new NegativeArraySizeException();
         }
-        if (len > MAX_LENGTH) {
+        if (len >= MAX_LENGTH) {
             throw new OutOfMemoryError("UTF16 String size is " + len +
-                                       ", should be less than " + MAX_LENGTH);
+                                       ", should be less than or equal to " + MAX_LENGTH);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/StringUTF16.java:59]:
```suggestion
                                       ", should be less than " + MAX_LENGTH);
```

Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/StringUTF16.java:59]: I second @ExE-Boss's comment.

Comment (RogerRiggs) [reply src/java.base/share/classes/java/lang/StringUTF16.java:59]: Please re-check the logic.
```
byte[] x = new byte[Integer.MAX_INTEGER];  
// throws java.lang.OutOfMemoryError: Requested array size exceeds VM limit
```
If `len  == StringUTF16.MAX_LENGTH`, then `len << 1` would be == Integer.MAX_VALUE.
And new byte[len << 1] would throw.

Note the change in the comparison from ">" to ">=".
The error message is describing the allowed range.

File: test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    private final static int MAX_UTF16_STRING_LENGTH = Integer.MAX_VALUE / 2;
+
+    private final static String EXPECTED_OOME_MESSAGE = "UTF16 String size is";
+
+    // Create a large UTF-8 byte array with a single non-latin1 character
+    private static byte[] generateUTF8Data(int byteSize) {
+        byte[] nonAscii = "\u0100".getBytes();
+        int nonAsciiSize = nonAscii.length;
+        int asciisize = byteSize - nonAsciiSize;
+        byte[] arr = new byte[asciisize + nonAsciiSize];
```
Comment (rgiulietti) [inline test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java:50]:
```suggestion
        byte[] arr = new byte[byteSize];
```
Note that `asciisize + nonAsciiSize` = `byteSize - nonAsciiSize + nonAsciiSize` = `byteSize`, even in the presence of overflows. Then `asciisize` becomes useless.

File: test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        Arrays.fill(arr, (byte)'x');      // fill with latin1
```
Comment (rgiulietti) [inline test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java:51]: `(byte) 0` is a valid UTF-8 sequence after all, so there is no need to fill the array with `'x'`.

File: test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        System.arraycopy(nonAscii, 0, arr, 0, nonAsciiSize); // non-latin1 at start
+        return arr;
+    }
+
+    // Create a large char array with a single non-latin1 character
+    private static char[] generateCharData(int size) {
+        char[] nonAscii = "\u0100".toCharArray();
+        int nonAsciiSize = nonAscii.length;
+        int asciisize = size - nonAsciiSize;
+        char[] arr = new char[asciisize + nonAsciiSize];
```
Comment (rgiulietti) [inline test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java:61]: Same as above

File: test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        Arrays.fill(arr, 'x');      // fill with latin1
```
Comment (rgiulietti) [inline test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java:62]: Same as above

**Comment by rgiulietti** [comment]
LGTM

**Review by rgiulietti — APPROVED** [review]

**Pull request closed without merging**